py3status/modules/speedtest.py

```diff
@@ -0,0 +1,273 @@
+        if any(
+            key.startswith(x)
+            for x in ["download", "upload"]
+            for key in self.placeholders
+        ):
+            if not any(key.startswith("upload") for key in self.placeholders):
+                self.command += " --no-upload"
+
+            if not any(key.startswith("download") for key in self.placeholders):
+                self.command += " --no-download"
+
+        if self.server_id and int(self.server_id):
+            self.command += " --server " + str(self.server_id)
+
+        if self.button_share and all(
+            x in self.placeholders for x in ["download", "upload"]
+        ):
+            self.command += " --share"
+
+        self.command += " --timeout " + str(self.timeout)
+
+    def _is_running(self):
+        try:
+            self.py3.command_output(["pgrep", "speedtest-cli"])
+            return True
+        except self.py3.CommandError:
+            return False
+
+    def _cast_number(self, value):
+        try:
+            value = int(value)
+        except ValueError:
+            try:
+                value = float(value)
+            except ValueError:
+                pass
+        return value
+
+    def _get_speedtest_data(self):
+        return loads(self.py3.command_output(self.command))
+
+    def _start(self):
+        self.cached_until = self.py3.time_in(self.sleep_timeout)
+
+        if not self._is_running():
+            self.cached_until = self.py3.CACHE_FOREVER
+            previous_data = self.py3.storage_get("speedtest_data") or {}
+            current_data = self._get_speedtest_data() or {}
+
+            if len(current_data) > 1:
+                # create a "total" for know if cnx is better or not
+                # between two run
+                current_data["total"] = int(current_data.get("download", 0)) + int(
+                    current_data.get("upload", 0)
+                )
+
+        # zero-ing if not fetched, raw version and units convertion
+        for x in ["download", "upload", "bytes_received", "bytes_sent"]:
+            unit = self.unit_size if "bytes" in x else self.unit_bitrate
+            current_data[x] = current_data.get(x, 0)
+            current_data[x + "_raw"] = current_data[x]
+            current_data[x], current_data[x + "_unit"] = self.py3.format_units(
+                current_data[x], unit=unit, si=self.si_units
+            )
+
+        # extra data, not sure we want to expose
+        current_data = self.py3.flatten_dict(
+            current_data, delimiter='_', intermediates=False
+        )
+
+        # store last data fetched
+        self.py3.storage_set("speedtest_data", current_data)
+
+        # get speedtest result url
+        self.url = current_data.get("share")
+
+        # create placeholders
+        self.speedtest_data.update(current_data)
+        self.speedtest_data.update(
+            {"previous_" + k: v for (k, v) in previous_data.items()}
+        )
```

Comment:
(Opinion) Are you going to use `_raw` and `previous_` yourself? If not, then I think it would be better not to add them in first place. Make it more about current speed test, not previous or raw. Users should be able to change `unit_bitrate` and `unit_size` to get same (raw) results too.

Reply:
raw was for do comparation, if now with patch and casting we can compare without raw, i can remove them.